pageserver/src/tenant/timeline.rs

```diff
@@ -1155,29 +1155,31 @@ impl Timeline {
         // FIXME One way around it would be not considering timeline as initialized before upload finishes.
         // So if there is no index file on the remote we consider all locally existing files as local only
         // and schedule their upload
-        // FIXME looks messy, needs cleanup
-        let (local_only_filenames, up_to_date_metadata) = if let Some(index_part) = index_part {
-            let local_only_filenames = self
-                .download_missing(index_part, remote_client, local_filenames, first_save)
-                .await?;
-            remote_client.init_upload_queue(index_part)?;
-            local_only_filenames
-        } else {
-            match remote_client.download_index_file().await {
+        let mut _downloaded_index_part = None;
+        let index_part: Result<&IndexPart, DownloadError> = match index_part {
+            Some(ip) => Ok(ip),
+            None => match remote_client.download_index_file().await {
                 Ok(index_part) => {
-                    let (local_only_filenames, up_to_date_metadata) = self
-                        .download_missing(&index_part, remote_client, local_filenames, first_save)
-                        .await?;
-                    remote_client.init_upload_queue(&index_part)?;
-                    (local_only_filenames, up_to_date_metadata)
-                }
-                Err(DownloadError::NotFound) => {
-                    info!("no index file was found on the remote");
-                    remote_client.init_upload_queue_empty(&local_metadata)?;
-                    (local_filenames, local_metadata)
+                    _downloaded_index_part = Some(index_part);
```

Comment:
In one of the intermediate steps I used Cow to combine reference with an owned value without temporary variables. May look a bit prettier, wdyt?

Reply:
Good idea, forgot that it exists. Will update this PR